ludwig/data/preprocessing.py

```diff
@@ -1205,6 +1241,55 @@ def build_data(
     return proc_cols
 
 
+def balance_data(dataset_df: DataFrame, output_features: List[Dict], preprocessing_parameters: Dict, backend: Backend):
+    """The purpose of this function is to balance the training dataset using either over-sampling or under-
+    sampling.
+
+    Args:
+        dataset_df: Input dataframe to be over-sampled or under-sampled.
+        output_features: List of feature configs.
+        preprocessing_parameters: Dictionary of the global preprocessing parameters.
+        backend: Backend for data processing.
+
+    Returns: An over-sampled or under-sampled training dataset.
+    """
+
+    if len(output_features) != 1:
+        raise ValueError("Class balancing is only available for datasets with a single output feature")
+    if output_features[0][TYPE] != BINARY:
+        raise ValueError("Class balancing is only supported for binary output types")
+
+    target = output_features[0][PROC_COLUMN]
+
+    if backend.df_engine.partitioned:
+        majority_class = backend.df_engine.compute(dataset_df[target].value_counts()).idxmax()
+        minority_class = backend.df_engine.compute(dataset_df[target].value_counts()).idxmin()
+    else:
+        majority_class = dataset_df[target].value_counts().idxmax()
+        minority_class = dataset_df[target].value_counts().idxmin()
+    majority_df = dataset_df[dataset_df[target] == majority_class]
+    minority_df = dataset_df[dataset_df[target] == minority_class]
+
+    if preprocessing_parameters["oversample_minority"] and preprocessing_parameters["undersample_majority"]:
+        raise ValueError(
+            "Cannot balance data if both oversampling an undersampling are specified in the config. "
+            "Must specify only one method"
+        )
+    elif preprocessing_parameters["oversample_minority"]:
+        sample_fraction = (len(majority_df) * preprocessing_parameters["oversample_minority"]) / len(minority_df)
+        minority_df = minority_df.sample(frac=sample_fraction, replace=True)
+    elif preprocessing_parameters["undersample_majority"]:
+        sample_fraction = int(len(minority_df) / preprocessing_parameters["undersample_majority"]) / len(majority_df)
+        majority_df = majority_df.sample(frac=sample_fraction, replace=False)
+
+    if backend.df_engine.partitioned:
+        balanced_df = backend.df_engine.concat([minority_df, majority_df])
+    else:
+        balanced_df = pd.concat([minority_df, majority_df])
```

Comment:
Wouldn't `backend.df_engine.concat` work even if the df engine was not partitioned?

Reply:
Yes actually it would. I think what happened here was I had named the backend concat function for dask concatenate which Justin pointed out was inconsistent so I changed it to concat, not realizing that it parallels the local backend setup. I'll change that now.